_delphi_utils_python/delphi_utils/validator/dynamic.py

```diff
@@ -461,89 +473,84 @@ def check_avg_val_vs_reference(self, df_to_test, df_to_reference, checking_date,
         Returns:
             - None
         """
-        # Average each of val, se, and sample_size over all dates for a given geo_id.
-        # Ignores NA by default.
-        df_to_test = df_to_test.groupby(['geo_id'], as_index=False)[
-            ['val', 'se', 'sample_size']].mean()
-        df_to_test["type"] = "test"
-
-        df_to_reference = df_to_reference.groupby(['geo_id'], as_index=False)[
-            ['val', 'se', 'sample_size']].mean()
-        df_to_reference["type"] = "reference"
-
-        df_all = pd.concat([df_to_test, df_to_reference])
+        # Calculate reference mean and standard deviation for each geo_id.
+        reference_mean = df_to_reference.groupby(['geo_id'], as_index=False)[
+            ['val', 'se', 'sample_size']].mean().assign(type="reference mean")
+        reference_sd = df_to_reference.groupby(['geo_id'], as_index=False)[
+            ['val', 'se', 'sample_size']].std().assign(type="reference sd")
+
+        # Replace standard deviations of 0 with non-zero min sd for that type. Ignores NA.
+        replacements = {"val": {0: df_to_test.val[df_to_test.val > 0].min()},
```

Comment:
Why do we need to replace mean-0 signals with a mean greater than 0?

Reply:
Whoops, the replacement values should've been taken from `reference_sd`. Thanks for the catch!

This is intended to replace the standard deviation of a zero-variance signal with the smallest non-zero standard deviation seen in another geo region of the same type (e.g. county) for the same signal (e.g. cases), since otherwise *any* change would trigger an error (z-score = inf with sd = 0).

Reply:
The other approach would be to just ignore/remove the zero-variance geo regions. Not sure if it makes a practical difference one way or the other. About 2% of counties have periods of zero-variance, based on a single spot-check.

Reply:
My question is specifically about the mean replacement. I don't know of any signals that could have a negative mean, but presumably it could be possible and the z-score can still be calculated and interpretable for negative means.

Reply:
Hm, I'm confused. Maybe there's a bug I'm not seeing, but this shouldn't be replacing mean values, only standard deviations.

[This bit above calculates standard deviations](https://github.com/cmu-delphi/covidcast-indicators/blob/24da6c0eec94fae80279dccca1bc3afc86628c59/_delphi_utils_python/delphi_utils/validator/dynamic.py#L479) for each of the `val`, `se`, and `sample_size` columns. So in the `reference_sd` dataframe, we end up with a column called `val` that is the standard deviation of the original `val` column (and `se` for sd of `se`, etc), it's not the mean (means are stored separately in `reference_mean`). The replacements defined here are [*only* being used on the `reference_sd` dataframe](https://github.com/cmu-delphi/covidcast-indicators/blob/24da6c0eec94fae80279dccca1bc3afc86628c59/_delphi_utils_python/delphi_utils/validator/dynamic.py#L486).

Reply:
Ah I understand now, I'm glad my confusion allowed you to catch a bug